File: src/qualtrics/reporting/static/flow.js
/* Accessible offline map and hypothetical walkthrough; all labels use textContent. */
(function (root) {
  'use strict';
  const clone = value => JSON.parse(JSON.stringify(value));
  const fresh = () => ({answers:{}, blockAnswers:{}, embedded:{}, choices:{}, completed:[]});
  const owns = (object, key) => Object.prototype.hasOwnProperty.call(object, key);

  function createController(document, payload, engine) {
    const $ = id => document.getElementById('flow-' + id);
    const all = selector => [...document.querySelectorAll(selector)];
    const surveys = payload.surveys || [];
    let currentSurvey = null, started = false, scenario = fresh(), history = [], result = null, readInputs = [];
    const node = (tag, className, text) => {
      const element = document.createElement(tag);
      if (className) element.className = className;
      if (text !== undefined) element.textContent = text;
      return element;
    };
    const addOption = (select, value, label) => {
      const option = node('option', '', label); option.value = String(value); select.append(option);
    };
    function selectControl(id, label, options, value = '') {
      const wrap = node('div', 'flow-input'), select = node('select'); select.id = id;
      const caption = node('label', '', label); caption.setAttribute('for', id);
      options.forEach(option => addOption(select, option.value, option.label)); select.value = value;
      wrap.append(caption, select); $('current').append(wrap); return select;
    }
    const findNode = (rootNode, id) => {
      if (rootNode.node_id === id) return rootNode;
      for (const child of rootNode.children || []) { const found = findNode(child, id); if (found) return found; }
      return null;
    };
    const titleFor = item => item ? currentSurvey.definition.blocks?.[item.config?.ID]?.name
      || item.config?.Description || ({Root:'Survey start', Block:'Question block', Standard:'Question block', Branch:'Branch', Group:'Group', Randomizer:'Randomizer', BlockRandomizer:'Randomizer', EndSurvey:'Survey ending', EmbeddedData:'Embedded data', WebService:'Web service', Authenticator:'Authenticator', Quota:'Quota', ReferenceSurvey:'Referenced survey'}[item.type]) || item.type : 'Survey flow';
    function reset() {
      started = false; scenario = fresh(); history = []; result = null; render();
    }
    function openAncestors(target, focus = false) {
      for (let ancestor = target; ancestor; ancestor = ancestor.parentElement) {
        if (ancestor.tagName === 'DETAILS') ancestor.open = true;
      }
      if (focus) { target.setAttribute('tabindex', '-1'); target.focus(); target.scrollIntoView({block:'nearest', behavior:'smooth'}); }
    }
    function markMap() {
      for (const survey of surveys) {
        for (const [id, anchor] of Object.entries(survey.nodes)) {
          const element = document.getElementById(anchor); if (!element) continue;
          const step = started && survey === currentSurvey ? result?.steps.find(item => item.node_id === id) : null;
          const pending = step?.status === 'pending' && result?.pending?.node_id === id;
          const state = !step ? 'configured' : pending ? 'pending' : step.status === 'pending' ? 'unreached' : step.status;
          const labels = {configured:'Configured', reached:'Reached', skipped:'Skipped', pending:'Pending', unreached:'Not reached'};
          element.dataset.flowStatus = state;
          const badge = element.querySelector('.flow-state'); if (badge) { badge.textContent = labels[state] || state; badge.title = step?.reason || ''; }
          if (pending) openAncestors(element);
        }
      }
    }
    function renderAnswer(input, index) {
      const occurrence = result.pending.node_id;
      const targetValues = draft => input.source === 'embedded' ? draft.embedded : (draft.blockAnswers[occurrence] ||= {});
      const values = input.source === 'embedded' ? scenario.embedded : (scenario.blockAnswers[occurrence] || {});
      const supplied = owns(values, input.key), value = values[input.key];
      const id = 'flow-answer-' + index, choices = input.choices || [];
      if (input.kind === 'choice') {
        const options = [{value:'unknown',label:'Not supplied (unknown)'}, {value:'blank',label:'No option selected (explicit answer)'},
          ...choices.map((choice, i) => ({value:String(i),label:choice.label}))];
        const selected = supplied ? (value === '' ? 'blank' : String(choices.findIndex(choice => choice.value === value))) : 'unknown';
        const control = selectControl(id, input.label, options, selected);
        readInputs.push(draft => {
          const target = targetValues(draft);
          if (control.value === 'unknown') delete target[input.key];
          else target[input.key] = control.value === 'blank' ? '' : choices[Number(control.value)]?.value;
        });
      } else if (input.kind === 'multiple') {
        const fieldset = node('fieldset', 'flow-input'); fieldset.append(node('legend', '', input.label));
        const selected = Array.isArray(value) ? value : [];
        const boxes = choices.map((choice, i) => {
          const label = node('label', 'flow-check'), box = node('input'); box.type = 'checkbox';box.id = id + '-' + i;
          box.checked = selected.includes(choice.value);label.append(box, node('span','',choice.label));fieldset.append(label);
          return {box, value:choice.value};
        });
        const empty = node('input');empty.type = 'checkbox';empty.checked = supplied && selected.length === 0;
        const blank = node('label','flow-check');blank.append(empty,node('span','','Explicitly no choices selected'));fieldset.append(blank);
        boxes.forEach(({box}) => box.addEventListener('change', () => { if (box.checked) empty.checked = false; }));
        empty.addEventListener('change', () => { if (empty.checked) boxes.forEach(({box}) => { box.checked = false; }); });
        $('current').append(fieldset);
        readInputs.push(draft => {
          const chosen = boxes.filter(({box}) => box.checked).map(item => item.value);
          const target = targetValues(draft);
          if (chosen.length || empty.checked) target[input.key] = chosen; else delete target[input.key];
        });
      } else {
        const wrap = node('div','flow-input'), caption = node('label','',input.label), control = node('input');
        caption.setAttribute('for',id);control.id=id;control.type=input.kind === 'number' ? 'number' : 'text';
        control.value=supplied ? String(value) : '';control.placeholder='Not supplied (unknown)';
        const explicit=node('input');explicit.type='checkbox';explicit.checked=supplied;
        const checkbox=node('label','flow-check');checkbox.append(explicit,node('span','','Use this value, including an empty value'));
        control.addEventListener('input',()=>{explicit.checked=true;});wrap.append(caption,control,checkbox);$('current').append(wrap);
        readInputs.push(draft=>{
          const target=targetValues(draft);
          if(explicit.checked) target[input.key]=control.value; else delete target[input.key];
        });
      }
    }
    function render() {
      readInputs=[]; $('current').replaceChildren(); $('error').hidden=true;
      $('continue').disabled=!currentSurvey; $('back').disabled=history.length === 0; $('reset').disabled=!started;
      $('continue').textContent=started ? 'Continue' : 'Start walkthrough';
      if (!currentSurvey) {
        $('walkthrough-status').textContent='Select a survey with an available flow definition to begin.';
        $('assumptions').hidden=true; markMap(); return;
      }
      if (!started) {
        $('walkthrough-status').textContent='Ready to explore ' + currentSurvey.label + '.';
        $('current').append(node('p','flow-note','Expand any map card to inspect its settings and questions. Start when you are ready to choose hypothetical answers.'));
        $('assumptions').hidden=true; markMap(); return;
      }
      result=engine.walkFlow(currentSurvey.definition,scenario);
      // Keep only reached occurrence answers and the current draft shown by Back.
      // A later appearance of the same question must have an independent answer.
      const pendingDraft=result.pending?.kind === 'block' ? scenario.blockAnswers[result.pending.node_id] : null;
      scenario.blockAnswers=clone(result.blockAnswers || {});
      if(pendingDraft) scenario.blockAnswers[result.pending.node_id]=pendingDraft;
      scenario.completed=[...(result.completed || [])];
      scenario.answers=clone(result.answers || {});
      const assumptions=[...(result.assumptions || [])];
      // Keep randomizer order visible even if the engine only records rule assumptions.
      for(const [id, order] of Object.entries(scenario.choices)) {
        if(Array.isArray(order) && result.steps.some(step=>step.node_id===id && step.status==='reached')) {
          assumptions.push('Randomizer ' + id + ': scenario order ' + order.map(child=>titleFor(findNode(currentSurvey.definition.root,child))).join(' → '));
        }
      }
      $('assumptions').hidden=assumptions.length === 0;
      $('assumption-list').replaceChildren(...assumptions.map(text=>node('li','',text)));
      markMap();
      const pending=result.pending;
      if(!pending) {
        $('continue').disabled=true;
        $('walkthrough-status').textContent=result.ended ? 'Scenario reached a survey ending.' : 'Scenario complete: no further configured steps.';
        $('current').append(node('p','flow-complete','Use Back to change earlier answers, or Reset to try another scenario.'));
        return;
      }
      const item=findNode(currentSurvey.definition.root,pending.node_id);
      const heading=node('h4','flow-current-title',titleFor(item));heading.setAttribute('tabindex','-1');
      const link=node('a','flow-current-link','Show this step in the map');link.href='#'+currentSurvey.nodes[pending.node_id];
      link.addEventListener('click',event=>{event.preventDefault();reveal(currentSurvey.nodes[pending.node_id]);});
      $('current').append(heading,link,node('p','flow-pending-reason',pending.reason));
      const completedCount=(result.completed || scenario.completed).length;
      $('walkthrough-status').textContent=`${completedCount} ${completedCount === 1 ? 'block' : 'blocks'} continued · ${pending.kind === 'block' ? 'Hypothetical answers' : 'Scenario choice required'}`;
      if(pending.canContinue === false) {
        $('continue').disabled=true;
        $('walkthrough-status').textContent='The scenario cannot continue with this definition.';
        $('current').append(node('p','flow-note','This step cannot continue. Inspect its definition, use Back to revise earlier choices, or Reset the scenario.'));
        return;
      }
      if(pending.kind === 'block') {
        (pending.inputs || []).forEach(renderAnswer);
        if(!pending.inputs?.length) $('current').append(node('p','flow-note','There are no supported answer inputs for this step. Continue to follow the configured route.'));
      } else if(pending.kind === 'condition') {
        const embeddedInputs=(pending.inputs || []).filter(input=>input.source === 'embedded');
        if(embeddedInputs.length) {
          $('current').append(node('p','flow-note','Supply hypothetical embedded values and recheck, or make an explicit condition assumption.'));
          embeddedInputs.forEach(renderAnswer);
          const recheck=node('button','','Recheck with these values');recheck.type='button';
          recheck.addEventListener('click',()=>{
            const draft=clone(scenario);readInputs.slice(0,embeddedInputs.length).forEach(read=>read(draft));
            history.push(clone(scenario));scenario=draft;render();
          });$('current').append(recheck);
        }
        $('current').append(node('p','flow-note','Unreached question answers cannot decide this condition. To explore further, label an assumption below.'));
        const select=selectControl('flow-condition-choice','Assume this condition',[
          {value:'',label:'Choose an explicit assumption'}, {value:'true',label:'Assume true — enter this branch'}, {value:'false',label:'Assume false — skip this branch'},
        ], typeof scenario.choices[pending.node_id] === 'boolean' ? String(scenario.choices[pending.node_id]) : '');
        readInputs.push(draft=>{
          if(select.value !== 'true' && select.value !== 'false') throw Error('Choose true or false explicitly before continuing.');
          draft.choices[pending.node_id]=select.value === 'true';
        });
      } else if(pending.kind === 'randomizer') {
        $('current').append(node('p','flow-note',`Choose ${pending.count} eligible ${pending.count === 1 ? 'step' : 'steps'} in the order this scenario should follow. This does not reproduce real allocation history.`));
        const controls=Array.from({length:Math.max(0,Math.min(500,pending.count || 0))},(_,index)=>selectControl('flow-draw-'+index,`Draw ${index+1}`,
          [{value:'',label:'Choose a step'},...(pending.options || [])],scenario.choices[pending.node_id]?.[index] || ''));
        readInputs.push(draft=>{
          const order=controls.map(control=>control.value);
          if(order.length !== pending.count || order.some(value=>!value) || new Set(order).size !== order.length) throw Error('Choose each required step once, in the order to follow.');
          draft.choices[pending.node_id]=order;
        });
      } else {
        const label=node('label','flow-check flow-assumption-choice'),check=node('input');check.type='checkbox';
        label.append(check,node('span','','For this scenario, assume this unsupported step allows continuation.'));
        $('current').append(node('p','flow-note','An explicit assumption is required. You can stop here and inspect the map without assuming a result.'),label);
        readInputs.push(draft=>{if(!check.checked) throw Error('Confirm the continuation assumption before proceeding.');draft.choices[pending.node_id]=true;});
      }
    }
    function proceed() {
      if(!currentSurvey) return;
      if(!started) { started=true;render();return; }
      if(!result?.pending || result.pending.canContinue === false) return;
      const draft=clone(scenario);
      try { readInputs.forEach(read=>read(draft)); }
      catch(error) { $('error').textContent=error.message;$('error').hidden=false;return; }
      const pending=result.pending;
      // Snapshot includes edits at this pending step, before advancing. Back
      // restores this exact prefix; later block answers and choices disappear.
      history.push(clone(pending.kind === 'block' ? draft : scenario));
      if(pending.kind === 'block') draft.completed=[...(result.completed || scenario.completed),pending.node_id];
      scenario=draft;render();
    }
    function update(selectedSurveyIds) {
      const selected=new Set(selectedSurveyIds), available=surveys.filter(survey=>selected.has(survey.id));
      all('.flow-survey').forEach(map=>{map.hidden=!selected.has(map.dataset.survey);});
      $('scope-empty').hidden=selected.size>0;
      const previous=currentSurvey?.id;
      $('survey-select').replaceChildren();available.forEach(survey=>addOption($('survey-select'),survey.id,survey.label));
      currentSurvey=available.find(survey=>survey.id===previous) || available[0] || null;
      $('survey-select').disabled=available.length<2; $('survey-select').value=currentSurvey?.id || '';
      if(currentSurvey?.id !== previous) reset();
    }
    function reveal(targetId) {
      const target=document.getElementById(String(targetId).replace(/^#/,''));
      if(!target || !all('.flow-node').includes(target)) return false;
      openAncestors(target,true);return true;
    }
    function records() {
      const rows=[];
      for(const survey of surveys) {
        function visit(item) {
          const target=document.getElementById(survey.nodes[item.node_id]);
          if(target) {
            const block=survey.definition.blocks?.[item.config?.ID];
            const questionText=(block?.elements || []).map(element=>survey.definition.questions?.[element.question_external_id]?.text || element.question_external_id || '').join(' ');
            const condition=item.config?.BranchLogic ? engine.describeCondition(item.config.BranchLogic,survey.definition) : '';
            const settings=(item.config?.EmbeddedData || []).map(entry=>[entry.Field,entry.Value,entry.Description].filter(value=>value !== undefined).join(' ')).join(' · ');
            rows.push({scope:'Flow',node:target,title:block?.name || item.config?.Description || item.type,
              content:[item.type,item.external_id,item.config?.ID,condition,questionText,settings].filter(Boolean).join(' · '),context:survey.label});
          }
          (item.children || []).forEach(visit);
        }
        visit(survey.definition.root);
      }
      return rows;
    }
    $('interactive').hidden=false;all('.flow-map-controls').forEach(control=>{control.hidden=false;});
    $('continue').addEventListener('click',proceed);
    $('back').addEventListener('click',()=>{if(history.length) {scenario=history.pop();render();}});
    $('reset').addEventListener('click',reset);
    $('survey-select').addEventListener('change',()=>{currentSurvey=surveys.find(survey=>survey.id===$('survey-select').value) || null;reset();});
    $('expand').addEventListener('click',()=>all('.flow-survey').filter(map=>!map.hidden).forEach(map=>map.querySelectorAll('details').forEach(detail=>{detail.open=true;})));
    $('collapse').addEventListener('click',()=>all('.flow-survey').filter(map=>!map.hidden).forEach(map=>map.querySelectorAll('.flow-node').forEach(detail=>{detail.open=false;})));
    render();
    return {update,reveal,records};
  }
  let controller;
  const api={createController,
    init() {
      if(controller) return controller;
      const element=root.document?.getElementById('flow-data');
      if(element && root.QualtricsFlowEngine) {
        try { controller=createController(root.document,JSON.parse(element.textContent),root.QualtricsFlowEngine); }
        catch(error) { const status=root.document.getElementById('flow-walkthrough-status');if(status) status.textContent='The walkthrough could not be initialized. The configured map remains available.'; }
      }
      return controller;
    },
    update(ids) { api.init()?.update(ids); },reveal(id) { return api.init()?.reveal(id) || false; },records() { return api.init()?.records() || []; },
  };
  if(typeof module === 'object' && module.exports) module.exports=api;
  else { root.ReportFlow=api;api.init(); }
})(typeof window !== 'undefined' ? window : globalThis);

File: src/qualtrics/reporting/flow.py
"""Offline survey flow map and walkthrough markup."""

from __future__ import annotations

import html
import json
import re
from typing import Any

from ..models import EntitySet
from .components.primitives import page_heading

_TYPES = {
    "Root": "Survey start",
    "Block": "Question block",
    "Standard": "Question block",
    "Branch": "Branch",
    "Group": "Group",
    "Randomizer": "Randomizer",
    "BlockRandomizer": "Randomizer",
    "EmbeddedData": "Embedded data",
    "EndSurvey": "Survey ending",
    "WebService": "Web service",
    "Authenticator": "Authenticator",
    "Quota": "Quota",
    "ReferenceSurvey": "Referenced survey",
}
_CONFIG = {"ID", "Description", "BranchLogic", "SubSet", "EvenPresentation", "EmbeddedData", "EndingType"}


def _escape(value: Any) -> str:
    return html.escape(str(value), quote=True)


def _safe_definition(raw: Any) -> dict[str, Any]:
    """Allowlist a stored envelope again before sending it to a report browser."""
    if not isinstance(raw, dict) or raw.get("schema_version") != 1:
        raise ValueError("Unsupported flow definition")
    count = 0
    seen: set[str] = set()

    def node(item: Any, depth: int = 0) -> dict[str, Any]:
        nonlocal count
        count += 1
        if count > 5000 or depth > 100 or not isinstance(item, dict):
            raise ValueError("Invalid flow tree")
        node_id = item.get("node_id")
        if not isinstance(node_id, str) or node_id in seen or not isinstance(item.get("type"), str):
            raise ValueError("Invalid flow occurrence")
        seen.add(node_id)
        children = item.get("children", [])
        config = item.get("config", {})
        if not isinstance(children, list) or not isinstance(config, dict):
            raise ValueError("Invalid flow node")
        safe = {key: value for key, value in config.items() if key in _CONFIG}
        if "EmbeddedData" in safe:
            entries = safe["EmbeddedData"]
            safe["EmbeddedData"] = (
                [
                    {key: value for key, value in entry.items() if key in {"Field", "Value", "Type", "Description"}}
                    for entry in entries
                    if isinstance(entry, dict)
                ]
                if isinstance(entries, list)
                else []
            )
        return {
            "node_id": node_id,
            "type": item["type"],
            "external_id": str(item.get("external_id") or ""),
            "config": safe,
            "children": [node(child, depth + 1) for child in children],
        }

    blocks = raw.get("blocks", {})
    questions = raw.get("questions", {})
    if not isinstance(blocks, dict) or not isinstance(questions, dict):
        raise ValueError("Invalid flow references")
    return {
        "schema_version": 1,
        "source_format": str(raw.get("source_format") or ""),
        "root": node(raw.get("root")),
        "blocks": {
            key: {
                "name": str(value.get("name") or key),
                "type": str(value.get("type") or ""),
                "elements": [
                    {k: v for k, v in element.items() if k in {"type", "question_external_id", "order"}}
                    for element in value.get("elements", [])
                    if isinstance(element, dict)
                ],
                "options": {
                    k: v
                    for k, v in value.get("options", {}).items()
                    if k in {"Looping", "RandomizeQuestions", "skip_logic"}
                },
            }
            for key, value in blocks.items()
            if isinstance(value, dict)
        },
        "questions": {
            key: {
                k: v
                for k, v in value.items()
                if k in {"text", "type", "selector", "choices", "choice_order", "display_logic", "skip_logic"}
            }
            for key, value in questions.items()
            if isinstance(value, dict)
        },
    }


def _condition_text(logic: Any, questions: dict[str, Any], depth: int = 0) -> str:
    if not isinstance(logic, dict) or depth > 30:
        return "Condition needs an explicit assumption in the walkthrough."
    if logic.get("LogicType") == "Question":
        qid = str(logic.get("QuestionID") or "Question")
        question = questions.get(qid, {})
        label = str(question.get("text") or qid)
        locator = re.fullmatch(r"q://[^/]+/SelectableChoice/([^/]+)", str(logic.get("ChoiceLocator") or ""))
        operator = logic.get("Operator")
        if locator and operator in {"Selected", "NotSelected"}:
            choice = question.get("choices", {}).get(locator[1], locator[1])
            return f"{label}: {choice} is {'not selected' if operator == 'NotSelected' else 'selected'}"
    if logic.get("LogicType") == "EmbeddedField":
        field = logic.get("LeftOperand") or logic.get("Field") or "Embedded field"
        operators = {
            "EqualTo": "equals",
            "NotEqualTo": "does not equal",
            "GreaterThan": "is greater than",
            "LessThan": "is less than",
            "GreaterThanOrEqual": "is at least",
            "LessThanOrEqual": "is at most",
        }
        operator = operators.get(str(logic.get("Operator")))
        if operator:
            return f"{field} {operator} {logic.get('RightOperand', '')}"
    terms = [logic[key] for key in sorted((key for key in logic if str(key).isdigit()), key=int)]
    if terms:
        parts = [_condition_text(term, questions, depth + 1) for term in terms]
        if len(parts) == 1:
            return parts[0]
        description = parts[0]
        for term, part in zip(terms[1:], parts[1:], strict=True):
            connector = term.get("Conjuction") or term.get("Conjunction") if isinstance(term, dict) else None
            conjunction = str(connector or logic.get("Conjunction") or logic.get("Operator") or "").lower()
            if conjunction not in {"and", "or"}:
                return "Conditions: " + "; ".join(parts) + ". Combination requires verification."
            description = f"({description}) {conjunction} ({part})"
        return description
    return "Condition needs an explicit assumption in the walkthrough."


def _render_node(
    item: dict[str, Any],
    definition: dict[str, Any],
    index: int,
    anchors: dict[str, str],
    targets: dict[str, str],
) -> str:
    nid = item["node_id"]
    anchor = f"flow-{index}-node-{len(anchors) + 1}"
    anchors[nid] = anchor
    kind = item["type"]
    config = item["config"]
    block = definition["blocks"].get(str(config.get("ID")), {})
    title = str(block.get("name") or config.get("Description") or _TYPES.get(kind, kind))
    settings = []
    if kind == "Branch":
        settings.append(
            "Continue into this branch when " + _condition_text(config.get("BranchLogic"), definition["questions"])
        )
    elif kind in {"Randomizer", "BlockRandomizer"}:
        count = config.get("SubSet", "the configured number of")
        noun = "step" if str(count) == "1" else "steps"
        settings.append(f"Select {count} eligible {noun} in an explicit scenario order.")
        if str(config.get("EvenPresentation", "")).lower() in {"true", "1"}:
            settings.append("Even presentation is configured; this offline scenario does not use allocation history.")
    elif kind == "EmbeddedData":
        for entry in config.get("EmbeddedData", []):
            field = entry.get("Field") or entry.get("Description") or "Unnamed field"
            settings.append(f"{field} = {entry['Value']}" if "Value" in entry else f"{field}: supplied at runtime")
    elif kind == "EndSurvey":
        settings.append("End the whole survey route here.")
        if config.get("EndingType"):
            settings.append(f"Ending: {config['EndingType']}")
    elif kind not in {"Root", "Block", "Standard", "Group"}:
        settings.append("This step needs an explicit assumption to proceed in the walkthrough.")
    for option, value in block.get("options", {}).items():
        if value and str(value).lower() not in {"false", "none", "0"}:
            settings.append(
                f"{'Question skip logic' if option == 'skip_logic' else option} is configured; "
                "the walkthrough may need an assumption."
            )
    questions = []
    for element in block.get("elements", []):
        if element.get("type") != "question":
            continue
        qid = str(element.get("question_external_id") or "")
        question = definition["questions"].get(qid, {})
        text = _escape(question.get("text") or qid or "Question")
        target = targets.get(qid)
        question_label = f"<a href='#{_escape(target)}'>{text}</a>" if target else text
        flags = "; ".join(
            name
            for key, name in (("display_logic", "Display logic"), ("skip_logic", "Skip logic"))
            if question.get(key)
        )
        questions.append(f"<li>{question_label}<small>{_escape(qid)}{' · ' + flags if flags else ''}</small></li>")
    question_html = (
        (
            f"<details class='flow-questions'><summary>{len(questions)} "
            f"{'question' if len(questions) == 1 else 'questions'} in this block</summary>"
            f"<ol>{''.join(questions)}</ol></details>"
        )
        if questions
        else ""
    )
    children = "".join(
        f"<li>{_render_node(child, definition, index, anchors, targets)}</li>" for child in item["children"]
    )
    children_html = f"<ol class='flow-children'>{children}</ol>" if children else ""
    identifiers = " · ".join(
        str(value) for value in (item.get("external_id"), config.get("ID"), f"Step {nid}") if value
    )
    return (
        f"<details id='{anchor}' class='flow-node' data-flow-node='{_escape(nid)}' open>"
        f"<summary><span class='flow-node-type'>{_escape(_TYPES.get(kind, kind))}</span>"
        f"<span class='flow-node-title'>{_escape(title)}</span><span class='flow-state'>Configured</span></summary>"
        f"<div class='flow-node-body'><p class='flow-node-ids'>{_escape(identifiers)}</p>"
        + "".join(f"<p class='flow-setting'>{_escape(setting)}</p>" for setting in settings)
        + question_html
        + children_html
        + "</div></details>"
    )


def render_flow(entities: EntitySet, question_targets: dict[tuple[str, str], str]) -> str:
    surveys: list[dict[str, Any]] = []
    maps = []
    for index, survey in enumerate(entities.surveys, 1):
        sid = str(survey["survey_id"])
        label = str(survey.get("survey_name") or sid)
        encoded = survey.get("flow_definition_json")
        definition = None
        message = "No flow definition is available. Re-parse with a survey definition that includes its flow."
        if encoded:
            try:
                definition = _safe_definition(json.loads(str(encoded)))
            except (ValueError, TypeError, RecursionError, AttributeError):
                message = "This flow definition could not be read. Re-parse it from the source definition."
        targets = {qid: target for (survey_id, qid), target in question_targets.items() if survey_id == sid}
        anchors: dict[str, str] = {}
        nodes = []
        if definition:
            nodes.append(_render_node(definition["root"], definition, index, anchors, targets))
            if not definition["root"]["children"]:
                nodes.append("<p class='flow-empty'>This flow definition contains no steps.</p>")
            surveys.append({"id": sid, "label": label, "definition": definition, "targets": targets, "nodes": anchors})
        else:
            nodes.append(f"<p class='flow-empty'>{message}</p>")
        maps.append(
            f"<section class='flow-survey' data-survey='{_escape(sid)}' aria-label='{_escape(label)} flow'>"
            f"<h3>{_escape(label)}</h3>{''.join(nodes)}</section>"
        )
    if not maps:
        maps.append("<p class='flow-empty'>No surveys are available.</p>")
    payload = json.dumps({"surveys": surveys}, ensure_ascii=True, separators=(",", ":")).replace("<", "\\u003c")
    return (
        "<section id='survey-flow' class='report-section report-view' aria-labelledby='flow-heading'>"
        + page_heading("Survey flow", heading_id="flow-heading")
        + "<p class='flow-intro'>Explore the configured route, then try hypothetical answers. "
        "This scenario uses the supplied definition; it does not show observed respondent routing "
        "or reproduce a Qualtrics preview.</p>"
        "<div class='flow-layout'><div class='flow-map'><div class='flow-map-head'><h3>Configured route</h3>"
        "<div class='flow-map-controls' hidden><button id='flow-outline-toggle' type='button' "
        "aria-expanded='false' aria-controls='flow-outline' hidden>Show outline</button></div></div>"
        "<div id='flow-canvas-shell' class='flow-canvas-shell' hidden>"
        "<div class='flow-canvas-toolbar'><h4 id='flow-canvas-title'>Configured route</h4>"
        "<div class='flow-zoom-controls'><button id='flow-zoom-out' type='button' aria-label='Zoom out'>−</button>"
        "<output id='flow-zoom-value' aria-label='Canvas zoom'>100%</output>"
        "<button id='flow-zoom-in' type='button' aria-label='Zoom in'>+</button>"
        "<button id='flow-fit' type='button'>Fit</button>"
        "<button id='flow-move-toggle' class='flow-move-toggle' type='button' aria-pressed='false'>Move map</button>"
        "<button id='flow-center' type='button'>Selected step</button></div></div>"
        "<p id='flow-canvas-help' class='flow-canvas-help'>Drag or scroll to pan · Ctrl/⌘ + scroll to zoom. "
        "Tab to a card to inspect it. Focus the canvas for arrow keys, +/− zoom, or Home to fit. "
        "On touch screens, swipe to scroll the page; choose Move map to pan, then Done moving to scroll again.</p>"
        "<div id='flow-canvas' class='flow-canvas' tabindex='0' role='region' "
        "aria-label='Survey route canvas' aria-describedby='flow-canvas-help'>"
        "<div id='flow-canvas-layer' class='flow-canvas-layer'></div></div>"
        "<p id='flow-canvas-empty' class='flow-empty' hidden>"
        "No flow definition is available for the selected scope.</p>"
        "</div>"
        "<div id='flow-outline' class='flow-outline'><div class='flow-outline-head'><h4>Accessible route outline</h4>"
        "<div class='flow-map-controls' hidden><button id='flow-expand' type='button'>Expand outline</button>"
        "<button id='flow-collapse' type='button'>Collapse outline</button></div></div>"
        "<p class='flow-legend'><span>Reached</span><span>Skipped</span>"
        "<span>Pending</span><span>Not reached</span></p>"
        "<p id='flow-scope-empty' class='flow-empty' hidden>Select a survey to see its flow.</p>"
        + "".join(maps)
        + "</div></div><aside class='flow-walkthrough' aria-labelledby='flow-walkthrough-heading'>"
        "<h3 id='flow-walkthrough-heading'>Explore this route</h3>"
        "<noscript>Enable JavaScript to use the walkthrough. The configured map remains readable.</noscript>"
        "<div id='flow-interactive' hidden><label for='flow-survey-select'>Map and walkthrough survey</label>"
        "<select id='flow-survey-select'></select>"
        "<div id='flow-panel-controls' class='flow-panel-controls' hidden aria-label='Flow panels'>"
        "<button id='flow-panel-details' type='button' aria-pressed='false' "
        "aria-controls='flow-selection'>Step details</button>"
        "<button id='flow-panel-scenario' type='button' aria-pressed='true' "
        "aria-controls='flow-scenario-panel'>What if</button>"
        "</div><section id='flow-selection' class='flow-selection' aria-label='Selected step details' hidden></section>"
        "<div id='flow-scenario-panel'><h4>Try a scenario</h4>"
        "<p class='flow-walkthrough-intro'>Your hypothetical answers stay in this page. "
        "Choose one survey and follow its route step by step.</p>"
        "<p id='flow-walkthrough-status' role='status' aria-live='polite'></p>"
        "<div id='flow-current'></div><p id='flow-error' role='alert' hidden></p>"
        "<div class='flow-actions'><button id='flow-back' type='button'>Back</button>"
        "<button id='flow-continue' type='button' class='flow-primary'>Start walkthrough</button>"
        "<button id='flow-reset' type='button'>Reset</button></div>"
        "<details id='flow-assumptions' class='flow-assumptions' hidden open><summary>Scenario assumptions</summary>"
        "<ol id='flow-assumption-list'></ol></details></div></div></aside></div></section>"
        f"<script id='flow-data' type='application/json'>{payload}</script>"
    )
